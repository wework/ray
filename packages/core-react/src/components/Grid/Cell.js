import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { settings } from '@wework/ray-core';

const breakpoints = Object.keys(settings.breakpointsInPixels);

export default function Cell({ className, push, span, ...props }) {
  const breakpointClasses = breakpoints.reduce((memo, breakpointName) => {
    const spanKey = `span-${breakpointName}`;
    if (props[spanKey]) {
      memo.push(`ray-grid__cell--span-${props[spanKey]}-${breakpointName}`);
    }

    const pushKey = `push-${breakpointName}`;
    if (props[pushKey]) {
      memo.push(`ray-grid__cell--push-${props[pushKey]}-${breakpointName}`);
    }

    return memo;
  }, []);

  const classes = clsx(
    'ray-grid__cell',
    {
      [`ray-grid__cell--span-${span}`]: span,
      [`ray-grid__cell--push-${push}`]: push
    },
    breakpointClasses,
    className
  );

  return <div className={classes} {...props} />;
}

Cell.propTypes = {
  className: PropTypes.string,
  span: PropTypes.number,
  push: PropTypes.number
};
