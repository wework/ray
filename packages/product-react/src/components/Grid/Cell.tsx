import clsx from 'clsx';
import React, { ReactHTML, ReactNode } from 'react';

export enum CellAlign {
  TOP = 'top',
  MIDDLE = 'middle',
  BOTTOM = 'bottom'
}

export enum CellDevice {
  DESKTOP = 'desktop',
  TABLET = 'tablet',
  PHONE = 'phone'
}

export enum CellJustify {
  TOP = 'top',
  MIDDLE = 'middle',
  BOTTOM = 'bottom'
}

export type CellProps = {
  align?: CellAlign;
  as: keyof ReactHTML;
  children?: ReactNode;
  className?: string;
  device?: CellDevice;
  full?: boolean;
  justify?: CellJustify;
  pushColumn?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
  span?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
} & typeof defaultProps;

const cellSpanClass = (span?: number, device?: CellDevice) => {
  if (span && device) {
    return 'ray-grid__cell--span-' + span + '-' + device;
  } else if (span && !device) {
    return 'ray-grid__cell--span-' + span;
  } else {
    return null;
  }
};

const cellAlignClass = (align?: CellAlign) => {
  if (align) {
    return 'ray-grid__cell--align-' + align;
  } else {
    return null;
  }
};

const cellPushClass = (pushColumn?: number, device?: CellDevice) => {
  if (pushColumn && device) {
    return 'ray-grid__cell--push-' + pushColumn + '-' + device;
  } else if (pushColumn && !device) {
    return 'ray-grid__cell--push-' + pushColumn;
  } else {
    return null;
  }
};

export const Cell = ({
  align,
  as: ComponentType,
  children,
  className,
  device,
  full,
  pushColumn,
  span,
  ...rest
}: CellProps) => (
  <ComponentType
    className={clsx(
      'ray-grid__cell',
      { 'ray-grid__cell--span-full': full },
      cellSpanClass(span, device),
      cellAlignClass(align),
      cellPushClass(pushColumn, device),
      className
    )}
    {...rest}
  >
    {children}
  </ComponentType>
);

Cell.CellAlign = CellAlign;
Cell.CellDevice = CellDevice;
Cell.CellJustify = CellJustify;

const defaultProps = {
  as: 'div'
};

Cell.defaultProps = defaultProps;

export default Cell;
