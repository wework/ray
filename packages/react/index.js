export { Container, Grid, Cell } from './src/components/Grid';
