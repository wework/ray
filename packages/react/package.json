{
  "name": "@wework/ray-react",
  "version": "0.0.1",
  "main": "index.js",
  "scripts": {
    "dev": "yarn storybook",
    "storybook": "start-storybook -p 6007 --ci"
  },
  "license": "Apache-2.0",
  "dependencies": {
    "@wework/ray-core": "1.3.1"
  },
  "devDependencies": {
    "@babel/core": "7.2.2",
    "@storybook/addon-knobs": "5.0.11",
    "@storybook/addon-options": "5.0.11",
    "@storybook/addon-viewport": "5.0.11",
    "@storybook/addons": "5.0.11",
    "@storybook/core": "5.0.11",
    "@storybook/react": "5.0.11",
    "babel-loader": "8.0.5",
    "clsx": "1.0.4",
    "prop-types": "15.7.2",
    "react": "16.8.6",
    "react-dom": "16.8.6",
    "storybook-chromatic": "1.3.3"
  },
  "peerDependencies": {
    "react": "^16.8.6",
    "react-dom": "^16.8.6"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "src/**/*"
  ],
  "keywords": [
    "wework",
    "react",
    "design-system"
  ]
}
