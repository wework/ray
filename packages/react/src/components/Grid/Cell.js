import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

const Cell = ({ children, className, span, push, ...props }) => {
  const keys = ['phone', 'tablet', 'desktop'];

  const spanObj = Array.isArray(span) ? span : [span, span, span];

  const {
    phone: spanPhone,
    tablet: spanTablet,
    desktop: spanDesktop
  } = spanObj.reduce((obj, val, i) => {
    const returnObj = obj;
    returnObj[keys[i]] = val;
    return returnObj;
  }, {});

  const pushObj = Array.isArray(push) ? push : [push, push, push];
  const {
    phone: pushPhone,
    tablet: pushTablet,
    desktop: pushDesktop
  } = pushObj.reduce((obj, val, i) => {
    const returnObj = obj;
    returnObj[keys[i]] = val;
    return returnObj;
  }, {});

  const classes = clsx({
    'ray-grid__cell': true,
    [`ray-grid__cell--span-${spanDesktop}-desktop`]: !!spanDesktop,
    [`ray-grid__cell--span-${spanTablet}-tablet`]: !!spanTablet,
    [`ray-grid__cell--span-${spanPhone}-phone`]: !!spanPhone,
    [`ray-grid__cell--push-${pushDesktop}-desktop`]: !!pushDesktop,
    [`ray-grid__cell--push-${pushTablet}-tablet`]: !!pushTablet,
    [`ray-grid__cell--push-${pushPhone}-phone`]: !!pushPhone,
    [`${className}`]: !!className
  });

  return (
    <div className={classes} {...props}>
      {children}
    </div>
  );
};

Cell.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  span: PropTypes.oneOfType([PropTypes.array, PropTypes.number]),
  push: PropTypes.oneOfType([PropTypes.array, PropTypes.number])
};

export default Cell;
