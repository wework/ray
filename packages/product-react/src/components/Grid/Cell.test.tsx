import { mount, ReactWrapper } from 'enzyme';
import * as React from 'react';
import { Cell } from './Cell';

describe('Cell', () => {
  let component: ReactWrapper<any>;

  beforeEach(() => {
    component = mount(<Cell />);
  });

  test('should match snapshot and styles', () => {
    expect(component).toMatchSnapshot();
  });

  test('should contain default class & passed in className', () => {
    component.setProps({ className: 'foo' });
    const container = component.find('div').at(0);
    expect(container.hasClass('ray-grid__cell')).toBe(true);
    expect(container.hasClass('foo')).toBe(true);
  });

  test('should accept "as" prop for setting element', () => {
    component.setProps({ as: 'main' });
    expect(component.find('div').length).toBe(0);
    expect(
      component
        .find('main')
        .at(0)
        .hasClass('ray-grid__cell'),
    ).toBe(true);
  });
  test('should pass additional props to component', () => {
    component.setProps({ id: 'foo', 'data-info': 'bar' });
    const container = component.find('.ray-grid__cell').at(0);
    expect(container.prop('id')).toEqual('foo');
    expect(container.prop('data-info')).toEqual('bar');
  });

  describe('default classes', () => {
    test('should contain default class', () => {
      expect(
        component
          .find('div')
          .at(0)
          .hasClass('ray-grid__cell'),
      ).toBe(true);
    });
  });

  describe('span classes', () => {
    test('should add class to span #', () => {
      component.setProps({ span: 1 });
      expect(
        component.find('.ray-grid__cell').hasClass('ray-grid__cell--span-1'),
      ).toBe(true);
      component.setProps({ span: 6 });
      expect(
        component.find('.ray-grid__cell').hasClass('ray-grid__cell--span-6'),
      ).toBe(true);
      component.setProps({ span: 12 });
      expect(
        component.find('.ray-grid__cell').hasClass('ray-grid__cell--span-12'),
      ).toBe(true);
    });
    test('should add class to span # w/ desktop device', () => {
      component.setProps({ span: 1, device: 'desktop' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--span-1-desktop'),
      ).toBe(true);
    });
    test('should add class to span # w/ desktop tablet', () => {
      component.setProps({ span: 6, device: 'tablet' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--span-6-tablet'),
      ).toBe(true);
    });
    test('should add class to span # w/ desktop phone', () => {
      component.setProps({ span: 12, device: 'phone' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--span-12-phone'),
      ).toBe(true);
    });
  });

  describe('align classes', () => {
    test('should add class to align top', () => {
      component.setProps({ align: 'top' });
      expect(
        component.find('.ray-grid__cell').hasClass('ray-grid__cell--align-top'),
      ).toBe(true);
    });
    test('should add class to align middle', () => {
      component.setProps({ align: 'middle' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--align-middle'),
      ).toBe(true);
    });
    test('should add class to align bottom', () => {
      component.setProps({ align: 'bottom' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--align-bottom'),
      ).toBe(true);
    });
  });

  describe('push classes', () => {
    test('should add class to push #', () => {
      component.setProps({ pushColumn: 1 });
      expect(
        component.find('.ray-grid__cell').hasClass('ray-grid__cell--push-1'),
      ).toBe(true);
      component.setProps({ pushColumn: 6 });
      expect(
        component.find('.ray-grid__cell').hasClass('ray-grid__cell--push-6'),
      ).toBe(true);
      component.setProps({ pushColumn: 12 });
      expect(
        component.find('.ray-grid__cell').hasClass('ray-grid__cell--push-12'),
      ).toBe(true);
    });
    test('should add class to push # w/ desktop device', () => {
      component.setProps({ pushColumn: 1, device: 'desktop' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--push-1-desktop'),
      ).toBe(true);
    });
    test('should add class to push # w/ desktop tablet', () => {
      component.setProps({ pushColumn: 6, device: 'tablet' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--push-6-tablet'),
      ).toBe(true);
    });
    test('should add class to push # w/ desktop phone', () => {
      component.setProps({ pushColumn: 12, device: 'phone' });
      expect(
        component
          .find('.ray-grid__cell')
          .hasClass('ray-grid__cell--push-12-phone'),
      ).toBe(true);
    });
  });
});