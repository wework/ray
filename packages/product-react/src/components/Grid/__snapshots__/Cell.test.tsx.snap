// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should match snapshot and styles 1`] = `
<Cell
  as="div"
>
  <div
    className="ray-grid__cell"
  />
</Cell>
`;
